Propagate reward tokens to descendent DAOs

jbaylina opened the pull request:
This PR tries to solve the problem with the Reward Tokens propagation in the split of a split DAO.

This implementation also allows reward tokens owners of a DAO to transfer his reward tokens to a different contract or to a regular account.

If this PR is merged, some tests should be done.

CJentzsch commented on DAO.sol:
```diff
@@ -715,6 +717,45 @@ contract DAO is DAOInterface, Token, TokenCreation {
+        return true;
+    }
+
+    function transferRewardTokens(address _to, uint _amountNum, uint _amountDen) external noEther returns (bool _success) {
```
with this you make rewardTokens freely tradeable, and DAO tokens value becomes very hard to distinguish.

CJentzsch commented on DAO.sol:
```diff
@@ -715,6 +717,45 @@ contract DAO is DAOInterface, Token, TokenCreation {
+            throw;
+        uint rewardTokensToTransfer = rewardToken[msg.sender] * _amountNum / _amountDen;
+
+        rewardToken[msg.sender] -= rewardTokensToTransfer;
```
this needs to be called by the motherDAO, why should they do it? Theres is no incentivization for it.

CJentzsch commented:
Not convinced that this solves the problem yet. Please convince me ;-)
If not, lets close it.

jbaylina commented:
transferRewardTokens is a recursive function. It transfer RT to a new address. But not only the RT of the self DAO, also the parent RTs owned by self, the grandparent RTs owned by self and so on. You can see that this function is called befor split to transfer the RT of self and all ancestors to the child DAO.

Reward tokens might be tradable now also. One way to do it is: If you set the curator and the DAO tokens owner of a child DAO to a single walled contract instead of an account. Then you just transfer/trade the ownership of the wallet.

If you want to limit, I don't understand exactly why, you can always limit this call to another contract.

collectInheritedDAOrewards is called by the child to get the rewards of the parent, grandparent and so on DAOs. This is the easy way for the child DAOs to collect all the rewards of all ancestors with a single call.

jbaylina commented:
I simplified and fixed a bug on this patch.  transferRewardTokens is not recursive any more and just transfers the reward tokens of this dao belonging to msg.self to another account

In the split process you transfer the reward tokens of the self DAO and all the reward tokens of all the ancestors that belongs to self.

collectInheritedDAOrewards is called by a child DAO to get the rewards of the parent, grandparent and all the ancestors DAOs. This is the easy way for the child DAOs to collect all the rewards of all ancestors with a single call.

CJentzsch commented:
cleaning up PRs (real solution to fair reward per project distribution is more complex)
